### ISSUE TITLE
Question: How to get line/rectangle in pdf file?

### ISSUE THREAD

**Issue by imckl** [pymupdf/PyMuPDF#382]
How to get line or rectangle in pdf file?
With lines/rectangles position, we are able to extract words from pdf more easily.
Like the image below (Red line/desc is what I want)
![image](https://user-images.githubusercontent.com/20377660/66548769-7bf68d00-eb74-11e9-8259-8efc2ab5098d.png)

**Comment by JorjMcKie** [comment]
Unfortunately, this is very difficult!
The reason is: you would have to write a syntax parser for a page's ``/Contents`` object, which can analyze the PDF mini-language used therein. See the PDF manual on page 985 ("Operator Summary").
A line is drawn with statements like
```
x y m    % move to point (x,y)
a b l    % draw line to point (a, b)
```
A rectangle can either be drawn by doing the previous for its 4 sides, or by
```python
x y width height re    % draw rect with bottom-left point (x,y)
```
All of the previous in **PDF** coordinates, which have to be translated to MuPDF coordinates ...
Even though PyMuPDF does have all the functions you need for doing this, I would still say: an almost hopeless endeavor, if you do not want to write your own PDF viewer.

I rather suggest you do a table analysis:
* Find the border of each column by analyzing the rectangles of the table header text. I assume this text is fairly fixed or well known, like ``Net. Weight``, ``Gross Weight``.
* Presumably, text extraction method ``getText("dict")`` is best for this, because it returns text pieces in the form of spans.
* Once you have the X0, X1 for e.g. ``Net. Weight``, then "5576" could be found by looking for words with coordinates x0 >= X0, x1 <= X1.
* Maybe you even **know** what those column limits are!

Some time ago I have developed a little interactive table extractor using GUI wxPython. Not sure if it still works correctly, but you might want to use things like [this](https://github.com/pymupdf/PyMuPDF/blob/master/examples/wxTableExtract.py).

**Comment by imckl** [comment]
Thanks for your reply.
- Currently I'm using some methods which similar to what you suggested, based on getText("rawdict"). But there are still cases hard to solve.
- In my case, the situation is more complicated than the pdf file chunk I pasted above. And that makes getText("dict")  is not accurate. So I have to do block work by hand from RAWDICT lol.
 1. this chunk, actually there are two different keywords, which treat as one block by getText("dict").
![image](https://user-images.githubusercontent.com/20377660/66567011-ea4d4680-eb98-11e9-83f3-65995e2d970a.png)
 2. and this chunk which words just break the normal table range (red line is 'column line' ), It's really annoying but it is the real world for what pdf files described.
![image](https://user-images.githubusercontent.com/20377660/66567565-11f0de80-eb9a-11e9-8178-eec056031498.png)
- I'm working on "key column" method, it's almost the right way to do such work but still has its difficulties. For example, there is a keyword with different names ( eg: net_weight <-> clear weight, Clean.Weight, Net., Net Weight(KG), and net weight split in two lines as net wei-, -ght, and even worse :( ). Again, It's really annoying but it is the real world for what pdf files described.
- I want to use rectangle border/ line as an auxiliary way to improve word recognition quality. However as you said, it might be a hard way to do in pdf. Therefore I should find another way.

**Comment by JorjMcKie** [comment]
I see. I'm glad I do not have to do your work ^^.
Maybe an approach could help, where you can use predefined column border coordinates. The formulars you are confronted with should fall into a small, maybe a handful of different ones. If that is true you would "only" have to find out which formular is the PDF you are currently working on, and then select the right set of vertical column coordinates.

**Comment by JorjMcKie** [comment]
Thought a little more about this interesting question, and developed a script that may be useful for you.
Don't expect too much: it is more a demo than anything else.
At least it seems to work good enough for your sample2.pdf. Have a look and use it as you like:

[sample2.zip](https://github.com/pymupdf/PyMuPDF/files/3719776/sample2.zip)

**Comment by imckl** [comment]
Thanks! I'll give it a try once have time :)

**Comment by imckl** [comment]
- Inspiration
I've tried your demo and it inspires me a lot, and now the rough idea of recognizing is to split multiple regions by keywords. Take this as an example:
![image](https://user-images.githubusercontent.com/20377660/67062054-cd1a0880-f194-11e9-8fae-2592af1430db.png)
let's say, BLUE - keyword, RED - region split by, GREEN - item. And this method meets up to half of the pdf files I've viewed. It will be a nice way to recognize if goes further.
- The hard part
Because the writing habit of ideogram (typically Chinese) is far different compared to Latin-ness(English and etc), it's hard to recognize PDF which mainly contains ideogram by default way provided by pymupdf.
I'm not a pro on linguistics, or deep learning or NLP or anything like that, but just currently working on some documents with the ideogram part. With the ease of pymupdf, I will try to find a better way to recognize ideograph words, and share my experience, and honestly, hoping this could contribute a little to pymupdf project :)
